src/tools/rbd/action/Group.cc

```diff
@@ -745,124 +769,1532 @@ int execute_group_snap_rollback(const po::variables_map &vm,
+{
+  int r;
+  string to;
+  // 4k limit to make sure we don't get a garbage string
+  r = group_read_string(fd, 4096, &to);
+  if (r < 0) {
+    std::cerr << "rbd: failed to decode end snap name" << std::endl;
+    return r;
+  }
+
+  bool exists;
+  r = group_snap_is_exist(rbd, group_ioctx, group_name, to, exists);
+  if (r < 0) {
+    return r;
+  }
+
+  if (exists) {
+    std::cerr << "end snapshot '" << to << "' already exists, aborting" << std::endl;
+    return -EEXIST;
+  }
+
+  tosnap = to;
+
+  return r;
+}
+
+static int group_import_snap_info(librbd::RBD &rbd, librados::IoCtx& group_ioctx,
+                        const char *group_name, int fd,
+                        string &to_snap, const int import_format)
+{
+  int r = utils::validate_banner(fd, utils::RBD_GROUP_BANNER_SNAP);
+  if (r < 0) {
+    return r;
+  }
+
+  while (r == 0) {
+    __u8 tag;
+    uint64_t length = 0;
+    r = utils::read_tag(fd, RBD_EXPORT_IMAGE_END, import_format, &tag, &length);
+    if (r < 0 || tag == RBD_EXPORT_IMAGE_END) {
+      break;
+    }
+
+    if (RBD_DIFF_FROM_SNAP == tag) {
+      r = do_group_snap_from(rbd, group_ioctx, group_name, fd);
+    } else if (RBD_DIFF_TO_SNAP == tag) {
+      r = do_group_image_snap_to(rbd, group_ioctx, group_name, fd, to_snap);
+    } else {
+      std::cerr << "rbd: invalid tag in image properties zone: " << tag
+                << ", Skip it." << std::endl;
+      r = utils::skip_tag(fd, length);
+    }
+  }
+
+  return r;
+}
+
+static int group_import_image_info(int fd,
+                        std::string &src_pool_name, std::string &src_image_name,
+                        uint64_t &is_create_image, const int import_format)
+{
+  int r = utils::validate_banner(fd, utils::RBD_GROUP_BANNER_IMAGE_INFO);
+  if (r < 0) {
+    return r;
+  }
+
+  while (r == 0) {
+    __u8 tag;
+    uint64_t length = 0;
+    r = utils::read_tag(fd, RBD_EXPORT_IMAGE_END, import_format, &tag, &length);
+    if (r < 0 || tag == RBD_EXPORT_IMAGE_END) {
+      break;
+    }
+
+    if (RBD_EXPORT_GROUP_IMAGE_POOL_NAME == tag) {
+      r = group_read_string(fd, 4096, &src_pool_name);
+    } else if (RBD_EXPORT_GROUP_IMAGE_NAME == tag) {
+      r = group_read_string(fd, 4096, &src_image_name);
+    } else if (RBD_EXPORT_GROUP_IS_CREATE_IMAGE == tag) {
+      r = group_decode_uint64(fd, is_create_image);
+    } else {
+      std::cerr << "rbd: invalid tag in image properties zone: "
+                << tag << ", Skip it." << std::endl;
+      r = utils::skip_tag(fd, length);
+    }
+  }
+
+  return r;
+}
+
+static int group_image_import_diff(librbd::RBD &rbd,
+                        librados::Rados &rados, librados::IoCtx& group_ioctx,
+                        const char *group_name, const string &to_snap, int fd,
+                        librbd::ImageOptions& opts, const size_t sparse_size,
+                        map<string, librbd::Image*> &images,
+                        vector<librbd::group_image_info_t> &images_list,
```

Comment:
nit: `image_list` seems to be `const`?

And `images` seems to be an output param, so should be passed via pointer. Also, why not use ` map<string, librbd::Image>`, so you don't have to free the pointer?

Reply:
`Image(const Image& rhs)`is private, so `map<string, librbd::Image>` cannot be used

Reply:
No, it just means you can't copy it. I suppose you was trying something like this:
```
Image image = ...;
images.insert({image_key, image});
```
So it would try to copy. But you could construct the image inside the map without copying and then use it by reference, e.g.:
```
Image &image = images.insert({image_key, {}}).first.second;
```
Or more readable:
```
auto &image = images[image_key];
```
it will automatically create the image entry if it does not exist.

Reply:
I have tried, the compilation failed, the code is as follows, please see if it is used like this:
```
map<string, librbd::Image> images_test;
std::string test_key("test");
librbd::Image &image_test = images_test.insert({test_key,{}}).first->second;
```

The failure message is as follows:
```
[root@localhost build]# ninja
[2/3] Building CXX object src/tools/rbd/CMakeFiles/rbd.dir/action/Group.cc.o
FAILED: src/tools/rbd/CMakeFiles/rbd.dir/action/Group.cc.o 
/usr/bin/c++  -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DHAVE_CONFIG_H -DHAVE_CURSES -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -isystem boost/include -Isrc/include -I../src -isystem include -isystem ../src/xxHash -isystem ../src/rapidjson/include -g -fPIE   -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wstrict-null-sentinel -Woverloaded-virtual -fno-new-ttp-matching -DCEPH_DEBUG_MUTEX -fstack-protector-strong -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -std=c++1z -MD -MT src/tools/rbd/CMakeFiles/rbd.dir/action/Group.cc.o -MF src/tools/rbd/CMakeFiles/rbd.dir/action/Group.cc.o.d -o src/tools/rbd/CMakeFiles/rbd.dir/action/Group.cc.o -c ../src/tools/rbd/action/Group.cc
../src/tools/rbd/action/Group.cc: In function ‘int rbd::action::group::group_import(librados::v14_2_0::Rados&, librados::v14_2_0::IoCtx&, const char*, int, librbd::ImageOptions&, size_t, bool, librbd::ProgressContext&)’:
../src/tools/rbd/action/Group.cc:1924:61: error: no matching function for call to ‘std::map<std::__cxx11::basic_string<char>, librbd::Image>::insert(<brace-enclosed initializer list>)’
   librbd::Image &image_test = images_test.insert({"test",{}}).first->second;
                                                             ^
In file included from /usr/include/c++/8/map:61,
                 from boost/include/boost/program_options/errors.hpp:15,
                 from boost/include/boost/program_options/options_description.hpp:12,
                 from boost/include/boost/program_options.hpp:15,
                 from ../src/tools/rbd/ArgumentTypes.h:12,
                 from ../src/tools/rbd/action/Group.cc:6:
/usr/include/c++/8/bits/stl_map.h:627:7: note: candidate: ‘std::map<_Key, _Tp, _Compare, _Alloc>::insert_return_type std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::map<_Key, _Tp, _Compare, _Alloc>::node_type&&) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::insert_return_type = std::_Node_insert_return<std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >, std::_Node_handle<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, librbd::Image>, std::allocator<std::_Rb_tree_node<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> > > > >; std::map<_Key, _Tp, _Compare, _Alloc>::node_type = std::_Node_handle<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, librbd::Image>, std::allocator<std::_Rb_tree_node<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> > > >]’
       insert(node_type&& __nh)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:627:7: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘std::map<std::__cxx11::basic_string<char>, librbd::Image>::node_type&&’ {aka ‘std::_Node_handle<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, librbd::Image>, std::allocator<std::_Rb_tree_node<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> > > >&&’}
/usr/include/c++/8/bits/stl_map.h:632:7: note: candidate: ‘std::map<_Key, _Tp, _Compare, _Alloc>::iterator std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator, std::map<_Key, _Tp, _Compare, _Alloc>::node_type&&) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::iterator = std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator = std::_Rb_tree_const_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::node_type = std::_Node_handle<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, librbd::Image>, std::allocator<std::_Rb_tree_node<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> > > >]’
       insert(const_iterator __hint, node_type&& __nh)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:632:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/8/bits/stl_map.h:801:7: note: candidate: ‘std::pair<typename std::_Rb_tree<_Key, std::pair<const _Key, _Tp>, std::_Select1st<std::pair<const _Key, _Tp> >, _Compare, typename __gnu_cxx::__alloc_traits<_Allocator>::rebind<std::pair<const _Key, _Tp> >::other>::iterator, bool> std::map<_Key, _Tp, _Compare, _Alloc>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; typename std::_Rb_tree<_Key, std::pair<const _Key, _Tp>, std::_Select1st<std::pair<const _Key, _Tp> >, _Compare, typename __gnu_cxx::__alloc_traits<_Allocator>::rebind<std::pair<const _Key, _Tp> >::other>::iterator = std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, librbd::Image>]’
       insert(const value_type& __x)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:801:7: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const value_type&’ {aka ‘const std::pair<const std::__cxx11::basic_string<char>, librbd::Image>&’}
/usr/include/c++/8/bits/stl_map.h:808:7: note: candidate: ‘std::pair<typename std::_Rb_tree<_Key, std::pair<const _Key, _Tp>, std::_Select1st<std::pair<const _Key, _Tp> >, _Compare, typename __gnu_cxx::__alloc_traits<_Allocator>::rebind<std::pair<const _Key, _Tp> >::other>::iterator, bool> std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::map<_Key, _Tp, _Compare, _Alloc>::value_type&&) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; typename std::_Rb_tree<_Key, std::pair<const _Key, _Tp>, std::_Select1st<std::pair<const _Key, _Tp> >, _Compare, typename __gnu_cxx::__alloc_traits<_Allocator>::rebind<std::pair<const _Key, _Tp> >::other>::iterator = std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, librbd::Image>]’
       insert(value_type&& __x)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:808:7: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘std::map<std::__cxx11::basic_string<char>, librbd::Image>::value_type&&’ {aka ‘std::pair<const std::__cxx11::basic_string<char>, librbd::Image>&&’}
/usr/include/c++/8/bits/stl_map.h:814:2: note: candidate: ‘template<class _Pair> std::__enable_if_t<std::is_constructible<std::pair<const _Key, _Tp>, _Pair>::value, std::pair<typename std::_Rb_tree<_Key, std::pair<const _Key, _Tp>, std::_Select1st<std::pair<const _Key, _Tp> >, _Compare, typename __gnu_cxx::__alloc_traits<_Allocator>::rebind<std::pair<const _Key, _Tp> >::other>::iterator, bool> > std::map<_Key, _Tp, _Compare, _Alloc>::insert(_Pair&&) [with _Pair = _Pair; _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >]’
  insert(_Pair&& __x)
  ^~~~~~
/usr/include/c++/8/bits/stl_map.h:814:2: note:   template argument deduction/substitution failed:
../src/tools/rbd/action/Group.cc:1924:61: note:   couldn't deduce template parameter ‘_Pair’
   librbd::Image &image_test = images_test.insert({"test",{}}).first->second;
                                                             ^
In file included from /usr/include/c++/8/map:61,
                 from boost/include/boost/program_options/errors.hpp:15,
                 from boost/include/boost/program_options/options_description.hpp:12,
                 from boost/include/boost/program_options.hpp:15,
                 from ../src/tools/rbd/ArgumentTypes.h:12,
                 from ../src/tools/rbd/action/Group.cc:6:
/usr/include/c++/8/bits/stl_map.h:828:7: note: candidate: ‘void std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::initializer_list<std::pair<const _Key, _Tp> >) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >]’
       insert(std::initializer_list<value_type> __list)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:828:7: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘std::initializer_list<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >’
/usr/include/c++/8/bits/stl_map.h:858:7: note: candidate: ‘std::map<_Key, _Tp, _Compare, _Alloc>::iterator std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator, const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::iterator = std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator = std::_Rb_tree_const_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, librbd::Image>]’
       insert(const_iterator __position, const value_type& __x)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:858:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/8/bits/stl_map.h:868:7: note: candidate: ‘std::map<_Key, _Tp, _Compare, _Alloc>::iterator std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator, std::map<_Key, _Tp, _Compare, _Alloc>::value_type&&) [with _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::iterator = std::_Rb_tree_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator = std::_Rb_tree_const_iterator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >; std::map<_Key, _Tp, _Compare, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, librbd::Image>]’
       insert(const_iterator __position, value_type&& __x)
       ^~~~~~
/usr/include/c++/8/bits/stl_map.h:868:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/8/bits/stl_map.h:873:2: note: candidate: ‘template<class _Pair> std::__enable_if_t<std::is_constructible<std::pair<const _Key, _Tp>, _Pair>::value, typename std::_Rb_tree<_Key, std::pair<const _Key, _Tp>, std::_Select1st<std::pair<const _Key, _Tp> >, _Compare, typename __gnu_cxx::__alloc_traits<_Allocator>::rebind<std::pair<const _Key, _Tp> >::other>::iterator> std::map<_Key, _Tp, _Compare, _Alloc>::insert(std::map<_Key, _Tp, _Compare, _Alloc>::const_iterator, _Pair&&) [with _Pair = _Pair; _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >]’
  insert(const_iterator __position, _Pair&& __x)
  ^~~~~~
/usr/include/c++/8/bits/stl_map.h:873:2: note:   template argument deduction/substitution failed:
../src/tools/rbd/action/Group.cc:1924:61: note:   candidate expects 2 arguments, 1 provided
   librbd::Image &image_test = images_test.insert({"test",{}}).first->second;
                                                             ^
In file included from /usr/include/c++/8/map:61,
                 from boost/include/boost/program_options/errors.hpp:15,
                 from boost/include/boost/program_options/options_description.hpp:12,
                 from boost/include/boost/program_options.hpp:15,
                 from ../src/tools/rbd/ArgumentTypes.h:12,
                 from ../src/tools/rbd/action/Group.cc:6:
/usr/include/c++/8/bits/stl_map.h:891:2: note: candidate: ‘template<class _InputIterator> void std::map<_Key, _Tp, _Compare, _Alloc>::insert(_InputIterator, _InputIterator) [with _InputIterator = _InputIterator; _Key = std::__cxx11::basic_string<char>; _Tp = librbd::Image; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, librbd::Image> >]’
  insert(_InputIterator __first, _InputIterator __last)
  ^~~~~~
/usr/include/c++/8/bits/stl_map.h:891:2: note:   template argument deduction/substitution failed:
../src/tools/rbd/action/Group.cc:1924:61: note:   candidate expects 2 arguments, 1 provided
   librbd::Image &image_test = images_test.insert({"test",{}}).first->second;
                                                             ^
../src/tools/rbd/action/Group.cc:1924:18: warning: unused variable ‘image_test’ [-Wunused-variable]
   librbd::Image &image_test = images_test.insert({"test",{}}).first->second;
                  ^~~~~~~~~~
ninja: build stopped: subcommand failed.
[root@localhost build]#
```

Reply:
Actually, I should have recommended `emplace` instead of `insert`, but this is still tricky.

You could use this instead though, which works for me:
```
map<string, librbd::Image> images_test;
std::string test_key("test");
auto &image_test = images_test[test_key];
```

Reply:
Thanks. That's a good idea.